tests/test_pex_builder.py

```diff
@@ -0,0 +1,200 @@
+        shutil.copytree(
+            repo_root / "tests/test-repos/dagster_project1",
+            tempdir_path / "dagster_project1",
+        )
+
+        import_pandas_code = tempdir_path / "imports_pandas.py"
+        import_pandas_code.write_text('import pandas\nprint("OK")\n')
+
+        # build with original deps
+        with run_builder(
+            builder_pex_path,
+            ["-m", "builder.deps", str(tempdir_path / "dagster_project1")],
+        ) as (build_output_dir, pex_files, other_files):
+            # one deps-HASH.pex is expected
+            assert 1 == len(pex_files)
+            deps_file = pex_files[0]
+
+            output = subprocess.check_output(
+                ["./" + deps_file, "-m", "dagster", "--version"],
+                cwd=build_output_dir,
+                stderr=subprocess.STDOUT,
+                encoding="utf-8",
+            )
+            assert "version" in output
+
+            # try to import nonexistent dependency
+            with pytest.raises(subprocess.CalledProcessError) as exc:
+                subprocess.check_output(
+                    ["./" + deps_file, str(import_pandas_code)],
+                    cwd=build_output_dir,
+                    stderr=subprocess.STDOUT,
+                    encoding="utf-8",
+                )
+            assert "No module named 'pandas'" in exc.value.output
+
+        # rebuild with new deps
+        (tempdir_path / "dagster_project1/requirements.txt").write_text("pandas\n")
+        with run_builder(
+            builder_pex_path,
+            ["-m", "builder.deps", str(tempdir_path / "dagster_project1")],
+        ) as (build_output_dir, pex_files, other_files):
+
+            deps_file = pex_files[0]
+            output = subprocess.check_output(
+                ["./" + deps_file, str(import_pandas_code)],
+                cwd=build_output_dir,
+                stderr=subprocess.STDOUT,
+                encoding="utf-8",
+            )
+            assert "OK" in output
+
+
+def test_builder_deploy_with_upload(builder_module, repo_root):
+    from builder import deploy
+
+    s3_objects = {}
+
+    def s3_urls_for_get(filenames):
+        return [
+            (filename if filename in s3_objects else None) for filename in filenames
+        ]
+
+    def s3_urls_for_put(filenames):
+        return filenames
+
+    def requests_get(url):
+        response = requests.Response()
+        if url in s3_objects:
+            response._content = s3_objects[url]
+            response.status_code = 200
+        else:
+            response.status_code = 404
+        return response
+
+    def requests_put(url, data):
+        s3_objects[url] = data.read() if hasattr(data, "read") else data
+        response = requests.Response()
+        response.status_code = 200
+        return response
+
+    dagster_project1_yaml = (
+        repo_root / "tests/test-repos/dagster_project1/dagster_cloud.yaml"
+    )
+    with mock.patch(
```

Comment:
https://github.com/getsentry/responses is a library with a nice api for mocking out requests if you're interested

Reply:
may switch as a follow up.